python/lsst/ip/diffim/detail/buildSingleKernelVisitor.cc

```diff
@@ -0,0 +1,98 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+#include <memory>
+#include <string>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <Eigen/Core>
+#include "numpy/arrayobject.h"
+#include "ndarray/pybind11.h"
+
+#include "lsst/afw/math/Kernel.h"
+#include "lsst/afw/math/SpatialCell.h"
+#include "lsst/ip/diffim/BuildSingleKernelVisitor.h"
+#include "lsst/pex/policy/Policy.h"
+
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+namespace lsst {
+namespace ip {
+namespace diffim {
+namespace detail {
+
+namespace {
+
+/**
+ * Wrap BuildSingleKernelVisitor and the factory function for one pixel type
+ *
+ * @tparam PixelT  Pixel type, e.g. `float`
+ * @param mod  pybind11 module
+ * @param[in] suffix  Class name suffix associated with PixeT, e.g. "F" for `float`
+ */
+template <typename PixelT>
+void declareBuildSingleKernelVisitor(py::module& mod, std::string const& suffix) {
+    py::class_<BuildSingleKernelVisitor<PixelT>, std::shared_ptr<BuildSingleKernelVisitor<PixelT>>,
+               afw::math::CandidateVisitor>
+        cls(mod, ("BuildSingleKernelVisitor" + suffix).c_str());
+
+    cls.def(py::init<afw::math::KernelList, pex::policy::Policy>(), "basisList"_a, "policy"_a);
+    cls.def(py::init<afw::math::KernelList, pex::policy::Policy, std::shared_ptr<Eigen::MatrixXd>>(),
+            "basisList"_a, "policy"_a, "hMat"_a);
+
+    cls.def("setSkipBuilt", &BuildSingleKernelVisitor<PixelT>::setSkipBuilt, "skip"_a);
+    cls.def("getNRejected", &BuildSingleKernelVisitor<PixelT>::getNRejected);
+    cls.def("getNProcessed", &BuildSingleKernelVisitor<PixelT>::getNProcessed);
+    cls.def("reset", &BuildSingleKernelVisitor<PixelT>::reset);
+    cls.def("processCandidate", &BuildSingleKernelVisitor<PixelT>::processCandidate, "candidate"_a);
+
+    mod.def("makeBuildSingleKernelVisitor", (std::shared_ptr<BuildSingleKernelVisitor<PixelT>>(*)(
+                                                afw::math::KernelList const&, pex::policy::Policy const&)) &
+                                                makeBuildSingleKernelVisitor<PixelT>,
+            "basisList"_a, "policy"_a);
+    mod.def("makeBuildSingleKernelVisitor",
+            (std::shared_ptr<BuildSingleKernelVisitor<PixelT>>(*)(
+                afw::math::KernelList const&, pex::policy::Policy const&, std::shared_ptr<Eigen::MatrixXd>)) &
```

Comment:
What precipitated the change is that one function that returned a `std::shared_ptr<Eigen::MatrixXd>` was failing in Python with a complaint that the returned value could not be converted to a Python object:
```
ipDiffim.makeRegularizationMatrix(self.policyDF)
TypeError: Unable to convert function return value to a Python type! The signature was
    (policy: lsst.pex.policy._policy.Policy) -> Eigen::Matrix<double, -1, -1, 0, -1, -1>
```
It was never clear to us why this was failing. It may be a clue that the function signature listed in the error message does not match the actual function signature, in that the shared pointer is missing. I am also unsure if the other functions that return shared pointer to matrices were working correctly. In any case...

Is was unwise to be using shared pointers to Eigen matrices, so rather than try to fix the pybind11 issue we decided to clean up the C++ API by removing use of shared pointers to Eigen matrices and vectors. Note that it would not have sufficed to fix just this one function because it produced data required by other functions, which required a shared pointer to a matrix. This also allowed us to improve const correctness.

@TallJimbo may wish to say more.